### PR TITLE
Improve indentation

### PR DESCRIPTION
**Pull request by fpringvaldsen**
This PR improves the indentation when modifying values in a ConfigDocument.

### CODE REVIEW

**Comment by fpringvaldsen** [comment]
Most of the significant quirks of this are called out in the tests, and it obviously won't work perfectly in the case where it gets a file with completely crazy formatting.

The only other quirk I can think of is the fact that, when indenting a setting on a value replacement, `ConfigNodeObject` uses the `indentation()` method, which finds the indentation of the first setting, so if there are other settings in the map that come before the one that matched and they have different indentations, the inserted value will be indented strangely. I'm going to work on fixing that now, as it shouldn't take long and shouldn't interfere with review of what's currently here.

**Comment by fpringvaldsen** [comment]
Okay, this should be good to go for review. @havocp Let me know if you want me to add more test coverage!

File: config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java [hunk 1]
```diff
@@ -24,4 +24,29 @@
...
+                childrenCopy.set(i, ((ConfigNodeComplexValue) childrenCopy.get(i)).indentText(indentation));
+            }
+        }
+        if (this instanceof ConfigNodeArray) {
+            return new ConfigNodeArray(childrenCopy);
+        } else if (this instanceof ConfigNodeObject) {
+            return new ConfigNodeObject(childrenCopy);
+        } else {
+            return new ConfigNodeConcatenation(childrenCopy);
+        }
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java]: This if/then/else feels like it should be a virtual method

File: config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java [hunk 2]
```diff
@@ -24,4 +24,29 @@
         }
         return tokens;
     }
+
+    protected ConfigNodeComplexValue indentText(AbstractConfigNode indentation) {
+        ArrayList<AbstractConfigNode> childrenCopy = new ArrayList<AbstractConfigNode>(children);
+        for (int i = 0; i < childrenCopy.size(); i++) {
+            if (childrenCopy.get(i) instanceof ConfigNodeSingleToken &&
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java]: if you do `AbstractConfigNode child = childrenCopy.get(i)` at the top here, you'd avoid a lot of noisy `childrenCopy.get(i)` repetition

**Comment by fpringvaldsen** [comment]
@havocp I've addressed your feedback! Also, I've noticed when it's searching for a setting from which to use the indentation, it won't look at include nodes, so I'm thinking about adding a test/functionality for that sometime later today.

**Comment by fpringvaldsen** [comment]
Also a question that came up while working through this: I've noticed that `ConfigNodeInclude` subclasses `ConfigNodeComplexValue`. Do we want it to work that way? Or should `ConfigNodeInclude` not be an instance of a value class?

**Comment by havocp** [comment]
I would like to factor meaningful semantic types of ConfigNode up into the public API as interfaces (ConfigNodeValue, ConfigNodeObject, ConfigNodeList, ConfigNodeInclude, ConfigNodeSubstitution, ...). Then the concrete types would be called Simple or Abstract or whatever. I was going to do this soon if you want.

**Comment by fpringvaldsen** [comment]
@havocp No rush! I think this PR gives us everything we need for our use case. I don't think changing `ConfigNodeInclude` to not subclass `ConfigNodeComplexValue` will change the way the config document stuff works. I was mostly just wondering if we should consider an include statement a "value". If so, the current implementation makes sense, but if not, changing `ConfigNodeInclude` to not subclass `ConfigNodeComplexValue` might make the implementation clearer.

**Comment by havocp** [comment]
if I remember right, includes expand into a bunch of fields that merge into the current object. so they are not really values.

**Comment by fpringvaldsen** [comment]
Thanks! I think it might make sense then to not subclass `ConfigNodeComplexValue` in `ConfigNodeInclude`, so I'll make a commit to change that.

**Comment by fpringvaldsen** [comment]
@havocp I've added that newest commit! I don't think there are any more changes I want to make on this PR, unless you have anymore feedback.

**Comment by havocp** [comment]
Thanks!

**Pull request merged**